skbase/tests/test_base.py

```diff
@@ -1036,3 +1039,60 @@ def test_has_implementation_of(
     # If the method is defined the first time in the parent class it should not
     # return _has_implementation_of == True
     assert not fixture_class_parent_instance._has_implementation_of("some_method")
+
+
+class FittableCompositionDummy(BaseEstimator):
+    """Potentially composite object, for testing."""
+
+    def __init__(self, foo, bar=84):
+        self.foo = foo
+        self.foo_ = deepcopy(foo)
+        self.bar = bar
+
+    def fit(self):
+        if hasattr(self.foo_, "fit"):
+            self.foo_.fit()
+        self._is_fitted = True
+
+
+def test_eq_dunder():
+    """Tests equality dunder for BaseObject descendants.
+
+    Equality should be determined only by get_params results.
+
+    Raises
+    ------
+    AssertionError if logic behind __eq__ is incorrect, logic tested:
+        equality of non-composites depends only on params, not on identity
+        equality of composites depends only on params, not on identity
+        result is not affected by fitting the estimator
+    """
+    non_composite = FittableCompositionDummy(foo=42)
+    non_composite_2 = FittableCompositionDummy(foo=42)
+    non_composite_3 = FittableCompositionDummy(foo=84)
+
+    composite = FittableCompositionDummy(foo=non_composite)
+    composite_2 = FittableCompositionDummy(foo=non_composite_2)
+    composite_3 = FittableCompositionDummy(foo=non_composite_3)
+
+    assert non_composite == non_composite
+    assert composite == composite
+    assert non_composite == non_composite_2
+    assert non_composite != non_composite_3
+    assert non_composite_2 != non_composite_3
+    assert composite == composite_2
+    assert composite != composite_3
+    assert composite_2 != composite_3
+
+    # equality should not be affected by fitting
+    composite.fit()
+    non_composite_2.fit()
+
+    assert non_composite == non_composite
+    assert composite == composite
```

Comment:
For these tests in general, can we reduce the number of asserts by taking advantage of the fact that if a == b, then we only need to test a == c and not b == c?  These are also really fast tests, so there isn't harm in leaving them.

Reply:
I think: no, and the reason is subtle.
I too saw the code quality comment.

The point is that we *do not already know* that equality as implemented is transitive, but *we need to test it* because we just implemented an equality that *should* satisfy that property.

I.e., theoretically, yes, it should be that a==c and a==b implies b==c - but we need to test whether actuality matches theory (rather than assume it).

Reply:
good point. If in first assert a==b, and in second assert a==c (or a!=c) we know the result for b==c (or b!=c). Likewise, if the first assert fails and a!=b, then when a==c (or a!=c), we know the outcome of b==c  if a==c, but if a!=c we don't know if b will (b/c it isn't equal to a).